Add secure update methods

mstemm opened the pull request:
@03cranec do you want to double-check the terminology I used for the split? check the first commit for details.

mstemm committed:
Split sdc into monitor and common

First step towards adding support for sysdig secure in the python sdc
client--splitting existing api into common and monitor-only
halves.

The new terminology is that "sdc" refers to the common code for both
monitor and secure. _SdcCommon is the (private) base class that has
endpoints/methods common to both monitor and secure. SdMonClient is the
class that provides all sysdig monitor capabilities, inheriting from
_SdcCommon. SdSecureClient will implement sysdig secure.

Split all the methods into the two classes as follows based on endpoint:

Common endpoints:
- /api/user/me
- /api/token (and /api/token/<user>/<id>)
- /api/agents/connected
- /api/notificationChannels
- /api/history/timelines
- /api/data
- /api/events
- /api/sysdig
- /api/users
- /api/teams
- /api/user
- /api/agents/config

Monitor-only endpoints:
- /api/data/metrics
- /api/alerts
- /api/notifications
- /data/drilldownDashboardDescriptors.json
- /data/drilldownDashboards
- /ui/dashboards
- /api/groupConfigurations
- /api/agents/falco_rules

Add __init__.py changes so all 3 objects are imported.

Change __checkResponse so it's simply hidden instead of hidden + mangled
with the class name. Allows sharing in inherited classes.

mstemm committed:
Add secure methods and tests.

Add methods
SdSecureClient.{get_set}_{user,system}_rules(new-rules-content). Set
actually only works for on-premise deployments. New examples
{get,set}_secure_{user,sytem}_falco_rules.py corespond to the
SdSecureClient methods.

Add new methods SdSecureClient.get_policy_events_{range,duration} to
fetch policy events. The API allows for an initial query based either on
duration or from/to. It returns a context dict containing the current
offset for pagination. That context can be provided to
get_more_policy_events() which increments the offset. The caller should
stop when the number of returned events is 0.

Also add a test script test/test_secure_apis.sh that can be used for
travis, and run that test script in travis. This script depends on a
staging account python-sdc-testing@draios.com. The API key and customer
id are passed to travis via environment variables.

The test script tests the following:

 - Trying to set the system falco rules file. This will fail with a
   405 (method not allowed) error, as this account is saas.
 - Trying to get the system falco rules file. This is compared to a
   known system rules file in test/sample-falco-rules.yaml.
 - Trying to set and get the user falco rules file. This uses the date
   as part of the rule name to ensure a fresh value is set/read.
 - Starting an agent with the python-sdc-testing account and triggering
   policy events, while checking for recent policy events. It should
   eventually find some policy events.

Also, for consistency's sake switch to external env variables for all
keys--instead of using a mix of embedded and externally provided
environment variables for api keys/access keys, use externally provided
keys everywhere.

philrz commented on README.md:
```diff
@@ -1,12 +1,12 @@
+A Python client API for Sysdig Monitor/Sysdig Secure.
 
-This module is a wrapper around the Sysdig Cloud API, which is documented [here](http://support.sysdigcloud.com/hc/en-us/articles/205233166-The-Sysdig-Cloud-API-Specification). It exposes most of the sysdig REST API functionality as an easy to use and easy to install Python interface. The repository includes a rich set of examples (in the [examples](examples/) subdir) that quickly address several ...
+This module is a wrapper around the Sysdig Monitor/Sysdig Secure APIs, which are documented [here](http://support.sysdigcloud.com/hc/en-us/articles/205233166-The-Sysdig-Cloud-API-Specification) and [here](xxx). It exposes most of the sysdig REST API functionality as an easy to use and easy to install Python interface. The repository includes a rich set of examples (in the [examples](examples/) su ...
```
Do you have a link to put in for that "xxx"?

mstemm replied:
Now that I know about readthedocs, I'll drop the second link and just point everything to the one page.

philrz commented on README.md:
```diff
@@ -20,36 +20,38 @@ Installation
+
 Once instantiated, all the methods documented below can be called on the object.
 
 ####Return Values
```
This wasn't your doing, but as long as you're in the code, might as well fix the fact that this "Return Values" bit doesn't render into a Markdown header.

mstemm replied:
Fixed.

philrz reviewed (commented):
tl;dr - Docs.

Background on docs: The current docs are generated via Sphinx using docstrings. What this gives us is that if you go to the `docs/` directory and type `make html`, it'll create docs with hyperlinks and all the trimmings (also options for PDF. etc.) And since we have that, it's also rigged up so that every time we commit something in GitHub, it triggers the recompile of the HTML docs and publishing them at http://python-sdc-client.readthedocs.io/en/latest/.

This leads to two things:

1. Ideally, the newly-added methods should have such docstrings. I know not every legacy method in Sysdig Monitor has them, but when I did the Sphinx conversation I at least brought all the semi-broken ones up to date. So it'd be good to be thorough going forward.

2. More importantly, splitting out the client into "common" and separate ones for Monitor/Secure has broken the current compilation of the docs. Unfortunately, while I got Sphinx working and was happy with it, I confess I am not a Sphinx *expert*. Its own docs confuse the hell out of me, and surely part of that is because while I can get by in Python, I am by no means an expert in Python in general.

Now, with a quick hack, it looks like the following diff at least gets the docs generating again:

```diff --git a/doc/index.rst b/doc/index.rst
index 9e27bb9..acdbdba 100644
--- a/doc/index.rst
+++ b/doc/index.rst
@@ -16,6 +16,11 @@ This page documents the functions available in the `Python Script Library <https
 Function List
 =============
 .. py:module:: sdcclient
-.. autoclass:: SdcClient
+.. autoclass:: SdMonClient
    :members:
+   :inherited-members:
+   :undoc-members:
+.. autoclass:: SdSecureClient
+   :members:
+   :inherited-members:
    :undoc-members:
```
What's maybe not great about what comes out of this is that it basically documents Monitor and Secure as separate whole animals as if the "common" methods they inherit belong to each, i.e. the page repeats the docstring for `get_user_info()`, etc. I did some quick searches to see how people deal with this problem, but nothing stood out. Maybe we just live with it. If you have some cycles to look deeper into it, that would be great.

mstemm committed:
Add methods to manage security policies

Add methods/example programs for limited management of security
policies:
 - create default policies from corresponding system falco rules file
 - list all policies
 - delete all policies

Add tests to use these example programs in conjunction with the system
rules file for the python-sdc-testing staging account.

mstemm committed:
Add sphinx docs

Add docs for all SdSecureClient methods. Also update the docs
specification to properly handle both the SdSecureClient and SdMonClient
objects (thanks @philrz!).

I was able to run "make docs" and browse the docs for both
objects. Shared methods for both objects are actually listed twice,
under each object. This seems ok for now as long as they're pulling docs
from a single location.

mstemm commented:
Ok, added sphinx docs and fixed the README. Wanna take a look again?

philrz commented:
Mmkay. Docs look good and automated tests are all passing, so, LGTM. :+1:
